File: 0085-maximal-rectangle/0085-maximal-rectangle.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int res = 0;
        vector<vector<int>> right(matrix.size(), vector<int>(matrix[0].size(), 0));
        vector<vector<int>> up(matrix.size(), vector<int>(matrix[0].size(), 0));
        vector<vector<int>> down(matrix.size(), vector<int>(matrix[0].size(), 0));
        for(int i = 0; i < matrix.size(); i++)
        {
            for(int j = matrix[0].size() - 1; j >= 0; j--)
            {
                if(matrix[i][j] == '1')
                {
                    if(j == matrix[0].size() - 1)
                    {
                        right[i][j] = 1;
                    }
                    else
                    {
                        right[i][j] = right[i][j + 1] + 1;
                    }
                }
            }
        }
        for(int j = 0; j < matrix[0].size(); j++)
        {
            stack<int> stk;
            if(right[0][j] > 0)
            {
                up[0][j] = 1;
                stk.push(0);
            }
            for(int i = 1; i < matrix.size(); i++)
            {
                if(right[i][j] == 0)
                {
                    while(!stk.empty())
                    {
                        stk.pop();
                    }
                    continue;
                }
                up[i][j] = 1;
                while(!stk.empty() && right[i][j] <= right[stk.top()][j])
                {
                    up[i][j] += up[stk.top()][j];
                    stk.pop();
                }
                stk.push(i);
            }
        }
        for(int j = 0; j < matrix[0].size(); j++)
        {
            stack<int> stk;
            if(right[matrix.size() - 1][j] > 0)
            {
                down[matrix.size() - 1][j] = 1;
                stk.push(matrix.size() - 1);
            }
            for(int i = matrix.size() - 2; i >= 0; i--)
            {
                if(right[i][j] == 0)
                {
                    while(!stk.empty())
                    {
                        stk.pop();
                    }
                    continue;
                }
                down[i][j] = 1;
                while(!stk.empty() && right[i][j] <= right[stk.top()][j])
                {
                    down[i][j] += down[stk.top()][j];
                    stk.pop();
                }
                stk.push(i);
            }
        }
        for(int i = 0; i < matrix.size(); i++)
        {
            for(int j = 0; j < matrix[0].size(); j++)
            {
                res = max(res, right[i][j] * (down[i][j] + up[i][j] - 1));
            }
        }
        return res;
    }
};

File: 0085-maximal-rectangle/0085-maximal-rectangle.py
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        res = 0
        right = [[0 for _ in range(len(matrix[0]))] for _ in range(len(matrix))]
        down = [[0 for _ in range(len(matrix[0]))] for _ in range(len(matrix))]
        up = [[0 for _ in range(len(matrix[0]))] for _ in range(len(matrix))]
        for i in range(len(matrix)):
            for j in range(len(matrix[0]) - 1, -1, -1):
                if matrix[i][j] == '1':
                    if j == len(matrix[0]) - 1:
                        right[i][j] = 1
                    else:
                        right[i][j] = right[i][j + 1] + 1
        for j in range(len(matrix[0])):
            stk = []
            for i in range(len(matrix)):
                while len(stk) > 0 and right[i][j] <= right[stk[-1]][j]:
                    stk.pop()
                if len(stk) == 0:
                    up[i][j] = i + 1
                else:
                    up[i][j] = i - stk[-1]
                stk.append(i)
            stk = []
            for i in range(len(matrix) - 1, -1, -1):
                while len(stk) > 0 and right[i][j] <= right[stk[-1]][j]:
                    stk.pop()
                if len(stk) == 0:
                    down[i][j] = len(matrix) - i
                else:
                    down[i][j] = stk[-1] - i
                stk.append(i)
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                res = max(res, right[i][j] * (up[i][j] + down[i][j] - 1))
        return res